### PR TITLE
Allow to configure how the events are delivered to the dispatchers.

### PR DESCRIPTION
**Pull request by armiol**

### CODE REVIEW

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL. Especially I am interested in your thoughts on naming for the new classes.

File: server/src/main/java/org/spine3/server/event/DispatcherEventPropagator.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ */
+@SPI
+@SuppressWarnings("WeakerAccess")   // Part of API.
+public abstract class DispatcherEventPropagator extends EventPropagator {
+
+    /**
+     * A pre-defined instance of the {@code DispatcherEventPropagator}, which does not postpone any event dispatching
+     * and uses {@link MoreExecutors#directExecutor()} for operation.
+     */
+    private static final DispatcherEventPropagator DIRECT_EXECUTOR = new DispatcherEventPropagator() {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/DispatcherEventPropagator.java]: Changed to a constant.

File: server/src/main/java/org/spine3/server/event/DispatcherEventPropagator.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    };
+
+    private Function<EventClass, Set<EventDispatcher>> dispatcherProvider;
+
+    @SuppressWarnings({"WeakerAccess", "unused"})   // Part of API.
+    protected DispatcherEventPropagator(Executor delegate) {
+        super(delegate);
+    }
+
+    @SuppressWarnings("WeakerAccess")   // Part of API.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/DispatcherEventPropagator.java]: Done.

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL.

That's still a work-in-progress. But I'd like to get your comments before this approach is extended to cover `EventSubscriber`s as well.

File: server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java [hunk 3]
```diff
@@ -34,34 +34,31 @@
 import static com.google.common.base.Preconditions.checkNotNull;
 
 /**
- * Propagates events from the {@link EventBus} to the matching {@link EventDispatcher}s.
+ * Executes the will of {@link EventBus} and delivers events to the matching {@link EventDispatcher}s.
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java]: > Executes the will of

It's a bit pathetic for Javadocs.

File: server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java [hunk 4]
```diff
@@ -150,4 +146,19 @@ public boolean apply(@Nullable EventDispatcher input) {
...
     }
+
+    /** Utility wrapper class for predefined executors designed to be constants */
+    private static class PredefinedExecutors {
+
+        /**
+         * A pre-defined instance of the {@code DispatcherEventExecutor}, which does not postpone any event dispatching
+         * and uses {@link MoreExecutors#directExecutor()} for operation.
+         */
+        private static final DispatcherEventExecutor DIRECT_EXECUTOR = new DispatcherEventExecutor() {
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java]: > `DIRECT_EXECUTOR`
This is not a constant. Have a look at Google guidelines as to naming such things.

File: server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java [hunk 5]
```diff
@@ -150,4 +146,19 @@ public boolean apply(@Nullable EventDispatcher input) {
...
+            @Override
+            public boolean maybePostponeDispatch(Event event, EventDispatcher dispatcher) {
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java]: > maybePostponeDispatch
How about just `postponeDispatch()`?

Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java]: In what circumstances an executer may return different values depending on content of parameters?

File: server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java [hunk 6]
```diff
@@ -150,4 +146,19 @@ public boolean apply(@Nullable EventDispatcher input) {
             }
         };
     }
+
+    /** Utility wrapper class for predefined executors designed to be constants */
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/DispatcherEventExecutor.java]: Please have a dot in the end of the sentence.

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL.

**Pull request merged**